### DISCUSSION TITLE
create a code-server app optimized for android

### DISCUSSION THREAD

**Discussion by ThiagoSilm** [coder/code-server#2840] [Ideas]
many people who use this project are on their cell phones for the simple fact that if we had a pc we would be using vscode itself, there is an app called vheditor that uses this project, but it is not very optimized, we will not be able to create a version of this project for android ? we would love it if that were possible or create an app with the same functionality as vs code.

**Comment by phisanti** [comment]
If this project gets an app, then programming on a DEX/EMUI phone will reach a whole new level. Let's mix it with some google cloud or paperspace computing power and full programming capabilities in the phone becomes a reality.

**Reply by jsjoeio** [reply]
I agree, this would be awesome! Technically, you could achieve this to some degree right now if you 
1. host your code-server in Google Cloud
2. add lots of computing power
3. access it from your mobile device

I guess the missing piece is "mobile-friendly" version of code-server for Android/iOS. It does feel a bit "clunky" using the mobile keyboard and trying to write code. Open to ideas!

**Reply by bpmct** [reply]
If you're using a mobile browser that supports it, you can install the PWA to your home screen. I was using `code-server --link` on my Macbook, but it could also be installed on a beefy Google Cloud server. 

https://user-images.githubusercontent.com/22407953/110371471-07e73f80-801b-11eb-8373-f8f97c08f4dc.mp4

I also connected an external keyboard to my USB C port on my Galaxy S9. I would agree that the internal keyboard is clunky. Ideally, connecting to an external monitor with something like Dex is the move.

**Comment by jsjoeio** [comment]
_Note: I've moved this over over to a discussion._ 

Woah! Hadn't seen vheditor before. What a cool project! I did use Termux the other day to install code-server on my Android. We have talked about creating specific mobile apps for code-server to optimize for developing from a phone.

> but it is not very optimized

What do you mean by this? What optimizations are you hoping for? 

> we would love it if that were possible or create an app with the same functionality as vs code.

By this do you mean having an Android app for develop from your phone with all the same functionality you get while developing in VS Code on a desktop? 

I would love to hear more about this and any thoughts or ideas you have.

**Reply by ThiagoSilm** [reply]
Well that's exactly what I wanted to say, there are no developers that are still starting or we are far from the computer, we can't do the things we do on the desktop, termux is a great tool, but it doesn't support frameworks (I say that from my own experience) like: react, react-native, vue and etc. it would be great to have a code editor for android like vscode, not with all the features but with the terminal and the editor working hahaha. this would help and make life easier for android developers, and you still wouldn't need to be on the desktop to continue your project

**Reply by jsjoeio** [reply]
I don't have bandwidth to take something on like this myself, but that couldn't change in the future. I know Termux is certainly a [hot topic](https://explodingtopics.com/topic/termux) so I imagine the need for something like this will continue to grow and in the future, we can reconsider and see if it aligns with our priorities.

**Reply by ThiagoSilm** [reply]
so would that be a yes?

**Reply by jsjoeio** [reply]
Lol that's a "yes, I think this is a good idea, but I don't have time to do this right now."

**Reply by ThiagoSilm** [reply]
this is good news, I hope you can get some time to put this idea into practice

**Reply by bpmct** [reply]
@Piposo Would you prefer for code-server and the code itself to run on your phone hardware?

If not, a free or low-cost [cloud service](https://github.com/cdr/deploy-code-server) can be used to host code-server and then it could just be accessed via the smart phone. In that case, I think this solution is possible today, especially with the PWA.

**Reply by ThiagoSilm** [reply]
@jsjoeio hey you couldn’t create an alpha version of vs code, with only the code editor included, and some things like deleting and creating new folders or files and then letting go to test ourselves and give us a hint?

**Reply by jsjoeio** [reply]
Sorry @Piposo I'm not following. Could you maybe rephrase what you mean? For playing with code-server, I suggest following the `CONTRIBUTING.md` guide [here](https://github.com/cdr/code-server/blob/main/docs/CONTRIBUTING.md).

**Reply by ThiagoSilm** [reply]
@jsjoeio well, I was trying to say that if you couldn’t make an alpha of the code server, a code app, very basic with the editor, so that you get excited about the design of an app haha

**Reply by jsjoeio** [reply]
Haha well I think the biggest question we need to answer is, what would the ideal editor experience look like on Android? Then see if there's a possibility of modifying code-server on Android to make it work that way

**Reply by bpmct** [reply]
> hey you couldn’t create an alpha version of vs code, with only the code editor included, and some things like deleting and creating new folders or files and then letting go to test ourselves and give us a hint

folders can be created / deleted / renamed from VS Code currently. are you suggesting just removing functionality from VS Code? why would this work better on android? run faster?

**Reply by ThiagoSilm** [reply]
Yes

**Reply by jsjoeio** [reply]
Following, I found this the other day which looks similar to what we're  discussing: [vhqtvn/VHEditor-Android](https://github.com/vhqtvn/VHEditor-Android)

**Reply by ThiagoSilm** [reply]
I had spoken of him, he is buggy

**Reply by jsjoeio** [reply]
Oh whoops, didn't realize that was the same vheditor you mentioned before. Gotcha :/ Well at least it's open source and we could draw some inspiration from it.

**Reply by ThiagoSilm** [reply]
This is very good, if what you are telling me is that you are going to make a vs code for android based on this project? I’m very happy if it’s a yes.

**Reply by jsjoeio** [reply]
Lol nope, just tossing out ideas as I see them. If/When we do this, you will know 😉

**Reply by ThiagoSilm** [reply]
Currently it is not possible to use vheditor because it is not possible to download nodejs, I am trying to implement an ubuntu terminal on it, as well as using the code server, to run npms that need permission administrator

**Reply by ThiagoSilm** [reply]
Take a look at this, I use this ubuntu terminal to run my projects  https://dev.to/josiasaurel/how-to-install-vscode-on-android-5f8d

**Reply by jsjoeio** [reply]
Ah, I see. Well at least that works, and Termux works so at least there are some solutions.

**Reply by ThiagoSilm** [reply]
You can implement it on vs code mobile, if it were implemented we would have a pc on the cell phone and this is real and possible

**Reply by ThiagoSilm** [reply]
well i'm excited about it, after all it would be a new era, now we could program away from the pc anywhere

**Reply by ThiagoSilm** [reply]
@jsjoeio I was thinking, couldn't you use that open source project based on code-server to make a better one than it?

**Reply by jsjoeio** [reply]
Theoretically, yes. If you want actually, you could open a new issue to summarize our discussion here and we could add it to the "Backlog Candidates" which would mean we wait for the community to weigh in and if there's enough requests for it + it makes strategically for us, we would consider doing it.

**Reply by ThiagoSilm** [reply]
can you explain better?

**Reply by jsjoeio** [reply]
Sure! I left a link on the other one. Can you open a new issue [here](https://github.com/cdr/code-server/issues/new?assignees=&labels=feature&template=feature-request.md&title=) summarizing and linking to this discussion? I'll handle the rest.

**Reply by ThiagoSilm** [reply]
I did what you said: https://github.com/cdr/code-server/issues/3370
